openstack/standalone-cinder/provisioner.go

```diff
@@ -0,0 +1,224 @@
+}
+
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	volumeId, err := createCinderVolume(p, options)
+	if err != nil {
+		glog.Errorf("Failed to create volume")
+		return nil, err
+	}
+
+	connection, err := connectCinderVolume(p, volumeId)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to connect volume: %v", err)
+		return nil, err
+	}
+
+	mapper, err := newVolumeMapperFromConnection(connection)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Unable to create volume mapper: %f" ,err)
+		return nil, err
+	}
+
+	ctx := provisionCtx{p, options, connection}
+	err = mapper.AuthSetup(ctx)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to prepare volume auth: %v", err)
+		return nil, err
+	}
+
+	pv, err := BuildPV(mapper, ctx, volumeId)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to build PV: %v", err)
+		return nil, err
+	}
+
+	return pv, nil
+}
+
+// Delete removes the storage asset that was created by Provision represented
+// by the given PV.
+func (p *cinderProvisioner) Delete(pv *v1.PersistentVolume) error {
+	ann, ok := pv.Annotations[provisionerIDAnn]
+	if !ok {
+		return errors.New("identity annotation not found on PV")
+	}
+	if ann != p.identity {
+		return &controller.IgnoredError{"identity annotation on PV does not match ours"}
+	}
+	// TODO when beta is removed, have to check kube version and pick v1/beta
+	// accordingly: maybe the controller lib should offer a function for that
+
+	volumeId, ok := pv.Annotations[cinderVolumeId]
+	if !ok {
+		return errors.New("cinder volume id annotation not found on PV")
+	}
+
+	ctx := deleteCtx{p, pv}
+	mapper, err := newVolumeMapperFromPV(ctx)
+	if err != nil {
+		return err
+	}
+
+	mapper.AuthTeardown(ctx)
+
+	err = disconnectCinderVolume(p, volumeId)
+	if err != nil {
+		return err
+	}
+
+	err = deleteCinderVolume(p, volumeId)
+	if err != nil {
+		return err
+	}
+
+	glog.Infof("Successfully deleted cinder volume %s", volumeId)
+	return nil
+}
+
+var (
+	master      = flag.String("master", "", "Master URL")
+	kubeconfig  = flag.String("kubeconfig", "", "Absolute path to the kubeconfig")
+	id          = flag.String("id", "", "Unique provisioner identity")
+	cloudconfig = flag.String("cloudconfig", "", "Path to OpenStack config file")
+)
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	if *cloudconfig == "" {
+		glog.Fatalf("missing OpenStack config file")
+	}
+
+	var config *rest.Config
+	var err error
+	if *master != "" || *kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags(*master, *kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+	prId := string(uuid.NewUUID())
+	if *id != "" {
```

Comment:
see https://github.com/kubernetes-incubator/external-storage/blob/master/ceph/rbd/cmd/rbd-provisioner/main.go#L67
```go
// By default, we use provisioner name as provisioner identity.
```

Reply:
Ok.  I'll change it to use the provisionerName